File: apps/ios/TaakHelden/Core/Parent/ParentModels.swift
import Foundation

// MARK: - Preview-layer models
// Local view-model layer for Phase 2 parent mode. Transport DTOs live in
// ContractModels / TaakHeldenAPIClient; this layer stays UI-friendly.

enum ParentSurface: String, CaseIterable, Identifiable {
    case vandaag
    case goedkeuren
    case taken
    case beloningen
    case instellingen

    var id: String { rawValue }
}

enum ParentTaskStatus: String, Codable, Equatable {
    case open
    case submitted
    case approved
    case completed
    case openRedo = "open_redo"
}

enum ParentTaskBucket: String, CaseIterable, Identifiable, Codable {
    case open
    case awaitingApproval
    case done

    var id: String { rawValue }

    var titleKey: String {
        switch self {
        case .open:
            return "parent.bucket.open"
        case .awaitingApproval:
            return "parent.bucket.awaiting"
        case .done:
            return "parent.bucket.done"
        }
    }
}

struct ParentPhotoAsset: Identifiable, Hashable, Equatable {
    let id: String
    let previewURL: URL?
    let accessibilityLabel: String
    var status: String?

    // Privacy by design: the fullscreen viewer never exposes EXIF or location.
    var showsSensitiveMetadata: Bool { false }
}

struct ParentTaskSnapshot: Identifiable, Equatable {
    let id: String
    let title: String
    let icon: String?
    let status: ParentTaskStatus
    let points: Int
    let submittedAt: Date?
    let photoAsset: ParentPhotoAsset?
    var photoStatus: String?

    var bucket: ParentTaskBucket {
        switch status {
        case .submitted:
            return .awaitingApproval
        case .approved, .completed:
            return .done
        case .open, .openRedo:
            return .open
        }
    }

    var statusLabelKey: String {
        switch status {
        case .open:
            return "parent.task.status.open"
        case .submitted:
            return "parent.task.status.submitted"
        case .approved:
            return "parent.task.status.approved"
        case .completed:
            return "parent.task.status.completed"
        case .openRedo:
            return "parent.task.status.redo"
        }
    }
}

struct ParentTodayChildSnapshot: Identifiable, Equatable {
    let id: String
    let displayName: String
    let avatar: String
    let balancePoints: Int
    let tasks: [ParentTaskSnapshot]

    var groupedTasks: [(bucket: ParentTaskBucket, items: [ParentTaskSnapshot])] {
        ParentTaskBucket.allCases.map { bucket in
            (bucket, tasks.filter { $0.bucket == bucket })
        }
    }
}

struct ApprovalQueueItem: Identifiable, Hashable, Equatable {
    let id: String
    let childID: String
    let childName: String
    let childAvatar: String
    let title: String
    let icon: String?
    let submittedAt: Date
    let points: Int
    let photoAsset: ParentPhotoAsset?
    var photoStatus: String?

    var hasPhoto: Bool { photoAsset != nil }
    var photoReady: Bool { photoStatus == "ready" }
    var photoProcessing: Bool { photoStatus == "processing" }
}

struct ApprovalQueueSection: Identifiable, Equatable {
    let id: String
    let childID: String
    let childName: String
    let childAvatar: String
    let items: [ApprovalQueueItem]
}

struct ParentManagedTask: Identifiable, Equatable {
    let id: String
    let title: String
    let icon: String?
    let points: Int
    let assigneeCount: Int
}

struct ParentManagedReward: Identifiable, Equatable {
    let id: String
    let title: String
    let icon: String?
    let price: Int
}

struct ParentSettingsSnapshot: Codable, Equatable {
    var soundEnabled: Bool
    let exportAvailable: Bool
    let deleteAvailable: Bool
}

struct ParentExportReceipt: Codable, Equatable {
    let message: String
}

struct ParentDashboardSnapshot: Equatable {
    let todayChildren: [ParentTodayChildSnapshot]
    let approvalSections: [ApprovalQueueSection]
    var managedTasks: [ParentManagedTask]
    var managedRewards: [ParentManagedReward]
    var settings: ParentSettingsSnapshot
    var lastSyncedAt: Date

    var pendingApprovalCount: Int {
        approvalSections.reduce(into: 0) { partialResult, section in
            partialResult += section.items.count
        }
    }

    var openTaskCount: Int {
        todayChildren.reduce(into: 0) { total, child in
            total += child.tasks.filter { $0.bucket == .open }.count
        }
    }

    func approvalItem(id: String) -> ApprovalQueueItem? {
        approvalSections.flatMap(\.items).first(where: { $0.id == id })
    }
}

enum BulkApprovalValidation: Equatable {
    case allowed
    case empty
    case mixedChildren
    case photoAcknowledgementRequired
}

enum ParentApprovalRules {
    static func validateBulkApproval(
        selectedItems: [ApprovalQueueItem],
        acknowledgedPhotoReview: Bool
    ) -> BulkApprovalValidation {
        guard !selectedItems.isEmpty else {
            return .empty
        }

        let childIDs = Set(selectedItems.map(\.childID))
        guard childIDs.count == 1 else {
            return .mixedChildren
        }

        if selectedItems.contains(where: \.hasPhoto), !acknowledgedPhotoReview {
            return .photoAcknowledgementRequired
        }

        return .allowed
    }
}

File: apps/android/core/src/main/kotlin/nl/taakhelden/core/parent/ParentModels.kt
package nl.taakhelden.core.parent

import java.time.Instant

/**
 * UI-friendly parent-mode models.
 *
 * Transport DTOs live in `nl.taakhelden.core.api`; this layer stays shaped for the
 * screens, exactly as the iOS `ParentModels.swift` does.
 */
public enum class ParentSurface {
    VANDAAG,
    GOEDKEUREN,
    TAKEN,
    BELONINGEN,
    INSTELLINGEN,
}

public enum class ParentTaskStatus(public val wire: String) {
    OPEN("open"),
    SUBMITTED("submitted"),
    APPROVED("approved"),
    COMPLETED("completed"),
    OPEN_REDO("open_redo"),
    ;

    public companion object {
        public fun fromWire(value: String?): ParentTaskStatus =
            entries.firstOrNull { it.wire == value } ?: OPEN
    }
}

public enum class ParentTaskBucket {
    OPEN,
    AWAITING_APPROVAL,
    DONE,
}

public data class ParentPhotoAsset(
    val id: String,
    val previewUrl: String? = null,
    val status: String? = null,
) {
    /** Privacy by design: the fullscreen viewer never exposes EXIF or location. */
    public val showsSensitiveMetadata: Boolean get() = false
}

public data class ParentTaskSnapshot(
    val id: String,
    val title: String,
    val icon: String?,
    val status: ParentTaskStatus,
    val points: Int,
    val submittedAt: Instant?,
    val photoAsset: ParentPhotoAsset?,
    val photoStatus: String? = null,
) {
    public val bucket: ParentTaskBucket
        get() = when (status) {
            ParentTaskStatus.SUBMITTED -> ParentTaskBucket.AWAITING_APPROVAL
            ParentTaskStatus.APPROVED, ParentTaskStatus.COMPLETED -> ParentTaskBucket.DONE
            ParentTaskStatus.OPEN, ParentTaskStatus.OPEN_REDO -> ParentTaskBucket.OPEN
        }
}

public data class ParentTodayChildSnapshot(
    val id: String,
    val displayName: String,
    val avatar: String,
    val balancePoints: Int,
    val tasks: List<ParentTaskSnapshot>,
) {
    /** Always emits all three buckets, so the columns stay stable when one is empty. */
    public val groupedTasks: List<Pair<ParentTaskBucket, List<ParentTaskSnapshot>>>
        get() = ParentTaskBucket.entries.map { bucket ->
            bucket to tasks.filter { it.bucket == bucket }
        }
}

public data class ApprovalQueueItem(
    val id: String,
    val childId: String,
    val childName: String,
    val childAvatar: String,
    val title: String,
    val icon: String?,
    val submittedAt: Instant,
    val points: Int,
    val photoAsset: ParentPhotoAsset?,
    val photoStatus: String? = null,
) {
    public val hasPhoto: Boolean get() = photoAsset != null
    public val photoReady: Boolean get() = photoStatus == "ready"
    public val photoProcessing: Boolean get() = photoStatus == "processing"
}

public data class ApprovalQueueSection(
    val id: String,
    val childId: String,
    val childName: String,
    val childAvatar: String,
    val items: List<ApprovalQueueItem>,
)

public data class ParentManagedTask(
    val id: String,
    val title: String,
    val icon: String?,
    val points: Int,
    val assigneeCount: Int,
)

public data class ParentManagedReward(
    val id: String,
    val title: String,
    val icon: String?,
    val price: Int,
)

public data class ParentSettingsSnapshot(
    val soundEnabled: Boolean,
    val exportAvailable: Boolean,
    val deleteAvailable: Boolean,
)

public data class ParentExportReceipt(
    val message: ExportReceiptMessage,
    val downloadUrl: String? = null,
)

/** Which export outcome to render; the copy itself lives in the string catalog. */
public enum class ExportReceiptMessage {
    READY,
    PENDING,
    FAILED,
}

public data class ParentDashboardSnapshot(
    val todayChildren: List<ParentTodayChildSnapshot>,
    val approvalSections: List<ApprovalQueueSection>,
    val managedTasks: List<ParentManagedTask>,
    val managedRewards: List<ParentManagedReward>,
    val settings: ParentSettingsSnapshot,
    val lastSyncedAt: Instant,
) {
    public val pendingApprovalCount: Int
        get() = approvalSections.sumOf { it.items.size }

    public val openTaskCount: Int
        get() = todayChildren.sumOf { child ->
            child.tasks.count { it.bucket == ParentTaskBucket.OPEN }
        }

    public fun approvalItem(id: String?): ApprovalQueueItem? {
        if (id == null) return null
        return approvalSections.asSequence().flatMap { it.items }.firstOrNull { it.id == id }
    }
}

public enum class BulkApprovalValidation {
    ALLOWED,
    EMPTY,
    MIXED_CHILDREN,
    PHOTO_ACKNOWLEDGEMENT_REQUIRED,
}

public object ParentApprovalRules {
    /**
     * Bulk approve is deliberately narrow:
     *  - one child at a time, so a parent never rubber-stamps across siblings;
     *  - photos need an explicit acknowledgement, because approving a photo unseen is
     *    the one action in the app a parent cannot take back.
     */
    public fun validateBulkApproval(
        selectedItems: List<ApprovalQueueItem>,
        acknowledgedPhotoReview: Boolean,
    ): BulkApprovalValidation {
        if (selectedItems.isEmpty()) return BulkApprovalValidation.EMPTY

        val childIds = selectedItems.map { it.childId }.toSet()
        if (childIds.size != 1) return BulkApprovalValidation.MIXED_CHILDREN

        if (selectedItems.any { it.hasPhoto } && !acknowledgedPhotoReview) {
            return BulkApprovalValidation.PHOTO_ACKNOWLEDGEMENT_REQUIRED
        }

        return BulkApprovalValidation.ALLOWED
    }
}
